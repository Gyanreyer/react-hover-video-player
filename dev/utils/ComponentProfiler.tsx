/* eslint-disable no-console */
import React from 'react';

interface RenderTiming {
  averageRenderTime: number;
  renderCount: number;
}

interface ComponentProfilerProps {
  profilerID: string;
  children: React.ReactChild;
}

/**
 * Util component logs out render timings for whatever component it's wrapped aroud
 */
const ComponentProfiler = ({
  profilerID,
  children,
}: ComponentProfilerProps): JSX.Element => {
  // Use a ref to track render timing for this component in a persistent store
  const renderTiming = React.useRef<RenderTiming>();
  if (!renderTiming.current) {
    renderTiming.current = {
      averageRenderTime: 0,
      renderCount: 0,
    };
  }

  // Logs out helpful render timing info for performance measurements
  const onProfilerRender = React.useCallback(
    (
      id, // the "id" prop of the Profiler tree that has just committed
      phase: string, // either "mount" (if the tree just mounted) or "update" (if it re-rendered)
      actualDuration: number // time spent rendering the committed update
    ) => {
      if (phase === 'mount') {
        console.log(`${profilerID} | MOUNT: ${actualDuration}ms`);
      } else {
        renderTiming.current.renderCount += 1;
        renderTiming.current.averageRenderTime +=
          (actualDuration - renderTiming.current.averageRenderTime) /
          renderTiming.current.renderCount;
        console.log(
          `${profilerID} | UPDATE: ${actualDuration}ms | New average: ${renderTiming.current.averageRenderTime}ms`
        );
      }
    },
    [profilerID]
  );
  return (
    <React.Profiler id={profilerID} onRender={onProfilerRender}>
      {children}
    </React.Profiler>
  );
};

export default ComponentProfiler;
