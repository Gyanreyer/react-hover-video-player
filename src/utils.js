/**
 * @typedef   {object}  VideoSource
 * @property  {string}  src - The src URL string to use for a video player source
 * @property  {string}  type - The media type of the video, ie 'video/mp4'
 */

/**
 * @function  formatVideoSrc
 *
 * Takes a videoSrc value and formats it as an array of VideoSource objects which can be used to render
 * <source> elements for the video
 *
 * @param {!(string|string[]|VideoSource|VideoSource[])}  videoSrc - Source(s) to use for the video player. Accepts 3 different formats:
 *                                                                   - **String**: the URL string to use as the video player's src
 *                                                                   - **Object**: an object with attributes:
 *                                                                     - src: The src URL string to use for a video player source
 *                                                                     - type: The media type of the video source, ie 'video/mp4'
 *                                                                   - **Array**: if you would like to provide multiple sources, you can provide an array of URL strings and/or objects with the shape described above
 *
 * @returns {VideoSource[]} Array of formatted VideoSource objects which can be used to render <source> elements for the video
 */
export function formatVideoSrc(videoSrc) {
  if (videoSrc == null) {
    // A videoSrc value is required in order to make the video player work
    console.error(
      "Error: 'videoSrc' prop is required for HoverVideoPlayer component"
    );

    return [];
  }

  return (
    // Make sure we can treat the videoSrc value as an array
    []
      .concat(videoSrc)
      // Parse our video source values into an array of VideoSource objects that can be used to render sources for the video
      .reduce((sourceArray, source) => {
        if (typeof source === 'string') {
          // If the source is a string, it's an src URL so format it into a VideoSource object and add it to the array
          sourceArray.push({ src: source });
        } else if (source && source.src) {
          // If the source is an object with an src, just add it to the array
          sourceArray.push({ src: source.src, type: source.type });
        } else {
          // Log an error if one of the videoSrc values is invalid
          console.error(
            "Error: invalid value provided to HoverVideoPlayer prop 'videoSrc':",
            source
          );
        }

        return sourceArray;
      }, [])
  );
}

/**
 * @typedef   {object}  VideoCaptionsTrack
 * @property  {string}  src - The src URL string for the captions track file
 * @property  {string}  srcLang - The language code for the language that these captions are in
 * @property  {string}  label - The title of the captions track
 * @property  {bool}    default - Whether this track should be used by default if the user's preferences don't match an available srcLang
 */

/**
 * @function formatVideoCaptions
 *
 * Takes a videoCaptions value and formats it as an array of VideoCaptionsTrack objects which can be used to render
 * <track> elements for the video
 *
 * @param {!(string|string[]|VideoCaptionsTrack|VideoCaptionsTrack[])} videoCaptions - Captions track(s) to use for the video player for accessibility. Accepts 3 different formats:
 *                                                                                     - **String**: the URL string to use as the captions track's src
 *                                                                                     - **Object**: an object with attributes:
 *                                                                                       - src: The src URL string for the captions track file
 *                                                                                       - srcLang: The language code for the language that these captions are in (ie, 'en', 'es', 'fr')
 *                                                                                       - label: The title of the captions track
 *                                                                                       - default: Whether this track should be used by default if the user's preferences don't match an available srcLang
 *                                                                                     - **Array**: if you would like to provide multiple caption tracks, you can provide an array of objects with the shape described above
 *
 * @returns {VideoCaptionsTrack[]}  Array of formatted VideoCaptionsTrack objects which can be used to render <track> elements for the video
 */
export function formatVideoCaptions(videoCaptions) {
  // If no captions were provided, return an empty array
  if (videoCaptions == null) return [];

  return (
    // Make sure we can treat the videoSrc value as an array
    []
      .concat(videoCaptions)
      // Parse our video captions values into an array of VideoCaptionsTrack
      // objects that can be used to render caption tracks for the video
      .reduce((captionsArray, captions) => {
        if (captions && captions.src) {
          captionsArray.push({
            src: captions.src,
            srcLang: captions.srcLang,
            label: captions.label,
            default: Boolean(captions.default),
          });
        } else {
          // Log an error if one of the videoCaptions values is invalid
          console.error(
            "Error: invalid value provided to HoverVideoPlayer prop 'videoCaptions'",
            captions
          );
        }

        return captionsArray;
      }, [])
  );
}
