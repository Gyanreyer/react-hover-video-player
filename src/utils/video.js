/* eslint-disable no-param-reassign */
// Enumerates states that the video can be in
export const VIDEO_STATE = {
  paused: 'paused',
  loading: 'loading',
  playing: 'playing',
};

/**
 * @function getVideoState
 *
 * Takes a video element and returns its current playing state
 *
 * @param {node} videoElement
 */
export function getVideoState(videoElement) {
  if (videoElement.paused || videoElement.ended) {
    return VIDEO_STATE.paused;
  }

  // If the video isn't paused but its readyState indicates it isn't loaded enough
  // to play yet, it is loading
  if (videoElement.readyState < 3) {
    return VIDEO_STATE.loading;
  }

  // If the video isn't paused and its ready state indicates it's loaded enough to play,
  // assume it's playing
  return VIDEO_STATE.playing;
}

/**
 * @function playVideo
 *
 * Takes a video element and attempts to play it.
 *
 * @param {node}  videoElement - The video element to start playing
 *
 * @returns {Promise}   A Promise which will resolve when the video starts playing or reject if an error occurs
 */
export function playVideo(videoElement) {
  // Start playing the video and hang onto the play promise it returns
  let playPromise = videoElement.play();

  if (!playPromise || !playPromise.then) {
    // If videoElement.play() didn't return a promise, we'll manually create one
    // ourselves which mimics the same behavior
    playPromise = new Promise((resolve, reject) => {
      // Declaring onVideoPlaybackFailed up here so we can refer to it and remove its event listener
      // if the video successfully starts playing
      let onVideoPlaybackFailed;

      // Set up event listener to resolve the promise when the video player starts playing
      const onVideoPlaybackStarted = () => {
        // Remove the event listeners we added as cleanup now that the play attempt has succeeded
        videoElement.removeEventListener('playing', onVideoPlaybackStarted);
        videoElement.removeEventListener('error', onVideoPlaybackFailed);

        // Resolve because we successfully started playing!
        resolve();
      };
      videoElement.addEventListener('playing', onVideoPlaybackStarted, {
        once: true,
      });

      // Set up event listener to reject the promise when the video player encounters an error
      onVideoPlaybackFailed = (event) => {
        // Remove the event listeners we added as cleanup now that the play attempt has failed
        videoElement.removeEventListener('error', onVideoPlaybackFailed);
        videoElement.removeEventListener('playing', onVideoPlaybackStarted);

        // Reject with the error that was thrown
        reject(event.error);
      };
      videoElement.addEventListener('error', onVideoPlaybackFailed);
    });
  }

  return playPromise;
}

/**
 * @function  pauseVideo
 *
 * Pauses the video and resets it to the beginning if necessary
 *
 * @param {node}  videoElement - The video element to pause playback for
 * @param {bool}  shouldRestart - Whether the videp should be reset to the beginning after pausing
 */
export function pauseVideo(videoElement, shouldRestart) {
  videoElement.pause();

  if (shouldRestart) {
    // If we should restart the video, reset its time to the beginning
    videoElement.currentTime = 0;
  }
}
