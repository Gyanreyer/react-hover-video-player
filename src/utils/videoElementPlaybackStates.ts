/**
 * Takes a video element and returns whether it is currently paused, meaning it is not
 * playing or attempting to play.
 *
 * @param {HTMLVideoElement} videoElement
 */
export function isVideoElementPaused(
  videoElement: HTMLVideoElement | null
): boolean {
  return videoElement ? videoElement.paused || videoElement.ended : true;
}
