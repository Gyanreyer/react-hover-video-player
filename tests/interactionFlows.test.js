import React from 'react';
import { fireEvent } from '@testing-library/react';

import { renderHoverVideoPlayer, advanceVideoTime } from './utils';

describe('Handles interactions correctly', () => {
  afterEach(() => {
    expect(console.error).not.toHaveBeenCalled();
  });

  test('mouseEnter and mouseLeave events take the video through its play and pause flows correctly', () => {
    const {
      videoElement,
      playerContainer,
      pausedOverlayWrapper,
      loadingOverlayWrapper,
    } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      loadingOverlay: <div />,
    });

    // The video should initially be paused
    expect(videoElement.play).not.toHaveBeenCalled();
    expect(videoElement).toBePaused();
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    // Advance time enough that the loading state timeout should have run but the video hasn't finished loading
    advanceVideoTime(200);

    // The loading overlay should now be visible
    expect(loadingOverlayWrapper).toBeVisible();

    // Advance time enough for the video to finish loading and play
    advanceVideoTime(200);
    expect(videoElement.play).toHaveBeenCalledTimes(1);

    expect(videoElement).toBePlaying();
    expect(pausedOverlayWrapper).not.toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Mouse out of the container to stop playing the video
    fireEvent.mouseLeave(playerContainer);

    // The video should not be paused until the timeout has completed
    expect(videoElement.pause).not.toHaveBeenCalled();

    // The paused overlay should be getting faded back in
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Advance time enough for the pause state transition to complete
    advanceVideoTime(400);

    expect(videoElement.pause).toHaveBeenCalledTimes(1);

    expect(videoElement).toBePaused();
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Mouse over one last time to start playing again
    fireEvent.mouseEnter(playerContainer);

    // The video should be playing right away since it's already loaded
    expect(videoElement).toBePlaying();
    expect(pausedOverlayWrapper).not.toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();
  });

  test('touch events take the video through its play and pause flows correctly', () => {
    const {
      videoElement,
      playerContainer,
      pausedOverlayWrapper,
      loadingOverlayWrapper,
    } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      loadingOverlay: <div />,
    });

    // The video should initially be paused
    expect(videoElement.play).not.toHaveBeenCalled();
    expect(videoElement).toBePaused();
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Touch the container to start playing the video
    fireEvent.touchStart(playerContainer);

    // Advance time enough that the loading state timeout should have run but the video hasn't finished loading
    advanceVideoTime(200);

    // The loading overlay should now be visible
    expect(loadingOverlayWrapper).toBeVisible();

    // Advance time enough for the video to finish loading and play
    advanceVideoTime(200);

    expect(videoElement).toBePlaying();
    expect(pausedOverlayWrapper).not.toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Touch somewhere outside of the element to stop "hovering"
    fireEvent.touchStart(document.body);

    // The video should not be paused until the timeout has completed
    expect(videoElement.pause).not.toHaveBeenCalled();

    // The paused overlay should be getting faded back in
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Advance time enough for the pause state transition to complete
    advanceVideoTime(400);

    expect(videoElement.pause).toHaveBeenCalledTimes(1);

    expect(videoElement).toBePaused();
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Touch the container one last time to start playing again
    fireEvent.touchStart(playerContainer);

    expect(videoElement).toBePlaying();
    expect(pausedOverlayWrapper).not.toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();
  });

  test('an attempt to play the video will correctly interrupt any attempts to pause it', () => {
    const {
      videoElement,
      playerContainer,
      pausedOverlayWrapper,
      loadingOverlayWrapper,
    } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      loadingOverlay: <div />,
    });

    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    // Tick forward so the video loads
    advanceVideoTime(300);

    expect(videoElement.play).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePlaying();

    // THe paused overlay should be hidden
    expect(pausedOverlayWrapper).not.toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // We are now in playing state, so mouse out to pause again
    fireEvent.mouseLeave(playerContainer);

    // At this point the video should have begun its pause attempt but not completed it
    expect(videoElement.pause).not.toHaveBeenCalled();
    expect(videoElement).toBePlaying();
    // The paused overlay should be fading back in
    expect(pausedOverlayWrapper).toBeVisible();

    // Mouse back over to cancel the pause attempt and start a play attempt
    fireEvent.mouseEnter(playerContainer);

    // The paused overlay should be hidden again
    expect(pausedOverlayWrapper).not.toBeVisible();

    // Advance time enough that the pause state transition timeout should have completed if it wasn't cancelled
    advanceVideoTime(300);

    // The video shouldn't have been paused
    expect(videoElement.pause).not.toHaveBeenCalled();
    expect(videoElement).toBePlaying();
    expect(pausedOverlayWrapper).not.toBeVisible();
  });

  test('an attempt to play the video when it is already loading will be handled correctly', () => {
    const {
      videoElement,
      playerContainer,
      pausedOverlayWrapper,
      loadingOverlayWrapper,
    } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      loadingOverlay: <div />,
      loadingStateTimeout: 50,
    });

    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    // Tick forward so the video is still loading
    advanceVideoTime(100);

    expect(videoElement.play).toHaveBeenCalledTimes(1);
    expect(videoElement).toBeLoading();

    // THe paused overlay should be hidden
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).toBeVisible();

    // We are now in playing state, so mouse out to pause again
    fireEvent.mouseLeave(playerContainer);

    // At this point the video should have begun its pause attempt but not completed it
    expect(videoElement.pause).not.toHaveBeenCalled();
    expect(videoElement).toBeLoading();
    // The paused overlay should be fading back in and the loading overlay should be hidden again
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Mouse back over to cancel the pause attempt and start a play attempt
    fireEvent.mouseEnter(playerContainer);

    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Tick forward so the loading overlay is shown again but the video still isn't done loading
    advanceVideoTime(100);

    expect(videoElement).toBeLoading();
    // The loading overlay should be visible
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).toBeVisible();

    // Allow the video to finish loading
    advanceVideoTime(100);

    expect(videoElement).toBePlaying();

    // The video shouldn't have been paused
    expect(videoElement.pause).not.toHaveBeenCalled();
    // Play should have only been called once
    expect(videoElement.play).toHaveBeenCalledTimes(1);
  });

  test('an attempt to play the video when it is already playing will be ignored', () => {
    const {
      videoElement,
      playerContainer,
      pausedOverlayWrapper,
      loadingOverlayWrapper,
    } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      loadingOverlay: <div />,
    });

    // Mouse over the container to start playing the video
    fireEvent.mouseEnter(playerContainer);

    // Tick forward so the video loads
    advanceVideoTime(300);

    expect(videoElement).toBePlaying();
    expect(pausedOverlayWrapper).not.toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // A second mouseEnter event should effectively be ignored
    fireEvent.mouseEnter(playerContainer);

    advanceVideoTime(300);

    // We should not have run through the play attempt flow again
    expect(videoElement.play).toHaveBeenCalledTimes(1);
    expect(videoElement).toBePlaying();
    expect(pausedOverlayWrapper).not.toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();
  });

  test('an attempt to pause the video when it is already paused will be ignored', () => {
    const {
      videoElement,
      playerContainer,
      pausedOverlayWrapper,
      loadingOverlayWrapper,
    } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      loadingOverlay: <div />,
    });

    // Mouse out of the container even though it was never started properly
    fireEvent.mouseLeave(playerContainer);

    // Advance time enough to prove that a pause attempt was not incorrectly started
    advanceVideoTime(300);

    expect(videoElement.pause).not.toHaveBeenCalled();
    expect(videoElement).toBePaused();
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();
  });

  test('pausing the video will correctly interrupt an active attempt to play the video', () => {
    const {
      videoElement,
      playerContainer,
      pausedOverlayWrapper,
      loadingOverlayWrapper,
    } = renderHoverVideoPlayer({
      videoSrc: 'fake/video-file.mp4',
      pausedOverlay: <div />,
      loadingOverlay: <div />,
      // Make the overlay transition duration fast enough that it won't disrupt
      overlayTransitionDuration: 50,
      loadingStateTimeout: 50,
    });

    expect(videoElement).toBePaused();

    // Start attempting to play the video
    fireEvent.mouseEnter(playerContainer);

    // Only allow the video to partially load
    advanceVideoTime(50);
    expect(videoElement).toBeLoading();
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).toBeVisible();

    // Mouse away from the video to cancel the play attempt
    fireEvent.mouseLeave(playerContainer);

    // The video is still in a loading state in the background, but the overlay state should have reverted to
    // only showing the paused overlay
    expect(videoElement).toBeLoading();
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();

    // Advance time enough so that the timeout to pause the video should run and not do anything since the play
    // promise still hasn't resolved
    advanceVideoTime(50);
    expect(videoElement).toBeLoading();

    // Advance time to allow the video's play attempt to resolve so the video can actually be paused
    advanceVideoTime(200);

    // The video should be paused
    expect(videoElement).toBePaused();
    expect(pausedOverlayWrapper).toBeVisible();
    expect(loadingOverlayWrapper).not.toBeVisible();
  });
});
