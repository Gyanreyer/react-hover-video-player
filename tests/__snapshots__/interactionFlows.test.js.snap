// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handles interactions correctly an attempt to pause the video when it is already paused will be ignored 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <div
      data-testid="paused-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 1; opacity: 1; transition: opacity 400ms; pointer-events: auto;"
    >
      <div />
    </div>
    <div
      data-testid="loading-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 2; opacity: 0; transition: opacity 400ms; pointer-events: none;"
    >
      <div />
    </div>
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles interactions correctly an attempt to play the video when it is already loading will be handled correctly 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <div
      data-testid="paused-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 1; opacity: 1; transition: opacity 400ms; pointer-events: auto;"
    >
      <div />
    </div>
    <div
      data-testid="loading-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 2; opacity: 0; transition: opacity 400ms; pointer-events: none;"
    >
      <div />
    </div>
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles interactions correctly an attempt to play the video when it is already playing will be ignored 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <div
      data-testid="paused-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 1; opacity: 1; transition: opacity 400ms; pointer-events: auto;"
    >
      <div />
    </div>
    <div
      data-testid="loading-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 2; opacity: 0; transition: opacity 400ms; pointer-events: none;"
    >
      <div />
    </div>
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles interactions correctly an attempt to play the video will correctly interrupt any attempts to pause it 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <div
      data-testid="paused-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 1; opacity: 1; transition: opacity 400ms; pointer-events: auto;"
    >
      <div />
    </div>
    <div
      data-testid="loading-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 2; opacity: 0; transition: opacity 400ms; pointer-events: none;"
    >
      <div />
    </div>
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles interactions correctly mouseEnter and mouseLeave events take the video through its play and pause flows correctly 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <div
      data-testid="paused-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 1; opacity: 1; transition: opacity 400ms; pointer-events: auto;"
    >
      <div />
    </div>
    <div
      data-testid="loading-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 2; opacity: 0; transition: opacity 400ms; pointer-events: none;"
    >
      <div />
    </div>
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles interactions correctly pausing the video will correctly interrupt an active attempt to play the video 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <div
      data-testid="paused-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 1; opacity: 1; transition: opacity 50ms; pointer-events: auto;"
    >
      <div />
    </div>
    <div
      data-testid="loading-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 2; opacity: 0; transition: opacity 50ms; pointer-events: none;"
    >
      <div />
    </div>
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;

exports[`Handles interactions correctly touch events take the video through its play and pause flows correctly 1`] = `
<div>
  <div
    data-testid="hover-video-player-container"
    style="position: relative;"
  >
    <div
      data-testid="paused-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 1; opacity: 1; transition: opacity 400ms; pointer-events: auto;"
    >
      <div />
    </div>
    <div
      data-testid="loading-overlay-wrapper"
      style="position: absolute; width: 100%; height: 100%; top: 0px; bottom: 0px; left: 0px; right: 0px; z-index: 2; opacity: 0; transition: opacity 400ms; pointer-events: none;"
    >
      <div />
    </div>
    <video
      crossorigin="anonymous"
      data-testid="video-element"
      loop=""
      playsinline=""
      style="display: block; width: 100%; object-fit: cover;"
    >
      <source
        src="fake/video-file.mp4"
      />
    </video>
  </div>
</div>
`;
